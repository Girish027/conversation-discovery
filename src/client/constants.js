import { basePath } from './swagger.json';
import { apiUrls } from './utils/apiUrls';

const domain = process.env.NODE_ENV === 'production' ? '//'.concat(window.location.host) : '//localhost:8080';

const constants = {
  environment: process.env.NODE_ENV,
  toolName: 'CDT',
  locale: 'en',
  eventNames: {
    logOutEvent: 'ClickLogoutUser',
    selectClientEvent: 'SelectClientEvent'
  },
  serverApiUrls: apiUrls,
  oktaApiUrls: {
    currentSession: '/api/v1/sessions/me'
  },
  udeFramework: 'https://d1af033869koo7.cloudfront.net/frontends/files/js/widget-loader-extensible.min.js',
  userOktaAccountLink: 'https://sso-247-inc.oktapreview.com/enduser/settings',
  navbar: {
    title: 'INTENT DISCOVERY',
    logoutUrl: domain.concat(`${basePath}/logout`)
  },
  contentUrl: {
    stable: 'http://virt-asst-content-stable.api.247-inc.net',
    production: 'http://virt-asst-content.api.247-inc.net',
    staging: 'http://virt-asst-content-staging.api.247-inc.net'
  },
  modals: {
    confirm: 'ConfirmationDialog',
    createProject: 'CreateProjectModal',
    editProject: 'EditProjectModal',
    createRun: 'CreateRunModal',
    editRun: 'EditRunModal',
    editCluster: 'EditClusterModal',
    progress: 'ProgressModal',
    unauthorized: 'SimpleModal',
    addToBot: 'AddToBotModal',
    addToFaq: 'AddToFaqModal'
  },

  buttons: {
    cancel: 'CANCEL',
    analyze: 'ANALYZE',
    analyzing: 'ANALYZING',
    submit: 'SUBMIT',
    save: 'SAVE',
    submitting: 'SUBMITTING...',
    next: 'NEXT',
    yesCancel: 'YES, CANCEL',
    dontCancel: 'NO, GO BACK',
    delete: 'DELETE',
    ok: 'OK',
    createNewRun: 'RUN A NEW ANALYSIS',
    createNewProject: 'CREATE A NEW PROJECT',
    uploadYourTranscripts: 'UPLOAD YOUR TRANSCRIPTS',
    start: 'START',
    refresh: 'REFRESH',
    download: 'DOWNLOAD',
    markAsComplete: 'MARK AS COMPLETE',
    viewRollupGraph: '<< VIEW ROLLUP DISTRIBUTION',
    viewGranularGraph: 'VIEW GRANULAR DISTRIBUTION',
    viewPieChart: 'VIEW DONUT CHART',
    viewBarChart: 'VIEW BAR CHART',
    runOverview: 'GO TO OVERVIEW',
    addToBot: 'ADD TO CONVERSATIONS',
    addToBotIntent: 'ADD',
    addingToBot: 'ADDING...',
    addToFAQ: 'ADD TO ANSWERS',
    expandAll: 'EXPAND ALL',
    collapseAll: 'COLLAPSE ALL',
    conversations: 'CONVERSATIONS',
    answers: 'ANSWERS'
  },
  status: {
    QUEUED: 'QUEUED',
    IN_PROGRESS: 'IN PROGRESS',
    ENABLED: 'ENABLED',
    DISABLED: 'DISABLED',
    VERIFYING: 'VERIFYING',
    FAILED: 'FAILED',
    READY: 'READY',
    COMPLETE: 'COMPLETE',
    ERROR: 'ERROR',
    NEW: 'NEW',
    IN_REVIEW: 'REVIEW',
  },
  projects: {
    SYSTEM: 'SYSTEM',
    MANUAL: 'MANUAL',
  },
  LIVE_DATA_ANALYSIS: 'Conversations Analysis',
  LIVE_DATA_IN_REVIEW: 'Conversations Analysis*',
  PROJECT_ALREADY_EXISTS: 'PROJECT_ALREADY_EXISTS',
  OTHER_CLUSTER: 'other',
  noop: () => { },
  maxDatasetFileSize: 200 * 1000000, // 200 MB
  pageTitle: 'Discover Intent',
  notification: {
    error: 'error',
    warning: 'warning',
    info: 'info',
    success: 'success'
  },
  pollingInterval: 10000,
  clustersPollingInterval: 30000,
  notificationInterval: 3000,

  // Modal Info
  modalInfo: {
    createRun: {
      header: 'Create a New Analysis',
      id: 'CreateRun'
    },
    editRun: {
      header: 'Edit Analysis Details',
      id: 'EditRun'
    },
    editProject: {
      header: 'Edit Transcript Details',
      id: 'EditProject'
    },
    editCluster: {
      header: 'Edit Topic',
      id: 'EditCluster'
    },
    addBot: {
      header: 'Add to Conversations',
      id: 'addBot'
    },
    addFaq: {
      header: 'Add to Answers',
      id: 'addFaq'
    }
  },
  utteranceMode: {
    options: ['Conversations', 'Answers'],
    answers: 'Answers',
    conversations: 'Conversations',
    modeLabel: 'MODE'
  },
  mappingTo: {
    intent: 'Intent',
    faq: 'FAQ'
  },

  // local storage keys
  localStorageBucket: 'cfdStore',
  activeProjectId: 'activeProjectId',
  isAnswers: 'isAnswers',
  isConversations: 'isConversations',
  userRole: 'userRole',
  selectedRunId: 'selectedRunId',
  clusterfilters: 'clusterfilters',
  clusterSort: 'clusterSort',
  clusterSearchString: 'clusterSearchString',

  sortOrder: {
    asc: 'asc',
    desc: 'desc',
    none: 'none',
  },

  // Distribution Graph
  DISTRIBUTION_GRAPH: {
    hAxis: 'Topic',
    vAxis: 'Volume',
    wordCloudTermsLable: 'wordCloudTerms',

    chartType: {
      barChart: 'ColumnChart',
      pieChart: 'PieChart',
    },
    graphOptions: {
      title: 'Topic Distribution',
      hAxis: { title: 'Topics' },
      vAxis: { title: 'Volume (Count)' },
      pieHole: 0.5,
      sliceVisibilityThreshold: 0.005,
    }
  },

  filterClusterByVolume: {
    volumeByCount: 'volumeByCount',
    volumeByPercent: 'volumeByPercent'
  },
  // Dialog Manager api defaluts
  nodeInfo: {
    parentNodeId: '00000000-0000-0000-0000-000000000000',
    responses: [{
      event: 'success', messageType: 'out.msg', mimeType: 'text/html', locale: 'en-US', offset: 0
    }]
  },
  previewWidgetPath: domain,
  ICON_DISCOVERY: 'icon_discovery',

  // Project Limit
  PROJ_LIMIT: 20,

  // Run Limit
  RUN_LIMIT: 20,
};

export const PREVIEW_ACTION_LIST = {
  conversationsInit: [
    'loadViewInchat',
  ],
};

export default constants;
