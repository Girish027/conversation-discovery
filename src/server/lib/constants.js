/*
  * 24/7 Customer, Inc. Confidential, Do Not Distribute. This is an
  * unpublished, proprietary work which is fully protected under
  * copyright law. This code may only be used pursuant to a valid
  * license from 24/7 Customer, Inc.
  */
'use strict';

module.exports = {
  /**
     * TODO: The name of the service
     * @type {string}
     */
  SERVICE_NAME: 'conversation-discovery',
  SCHEDULER_NAME: 'conversation-fetcher',

  /**
     * Disabled header
     * @type {string}
     * @private
     */
  X_POWERED_BY: 'x-powered-by',

  /**
     * Header that reflects the active version of the service
     * Used by the health check
     * @type {string}
     * @private
     */
  X_SERVICE_VERSION_HEADER: 'X-TFS-Service-Version',

  /**
   * Header to identify internal proxy calls
   * @type {string}
   * @private
   */
  X_TFS_INTERNALPROXYCALL: 'x-tfs-internalproxycall',

  /**
     * Common MIME types
     * @type {string}
     */
  MIME_APP_URLENCODED: 'application/x-www-form-urlencoded',
  MIME_APP_JSON: 'application/json',
  MIME_APP_XML: 'application/xml',
  MIME_TEXT_PLAIN: 'text/plain',

  /**
     * Config Keys
     * @type {string}
     */
  LOG_CONFIG_KEY: 'service-log.config-path',
  GC_LOG_ENABLED_CONFIG_KEY: 'gc-log.enabled',
  LISTEN_PORT_CONFIG_KEY: 'listen.port',
  KEEPALIVE_TIMEOUT_CONFIG_KEY: 'listen.keepAliveTimeout',
  REQUEST_LOG_CONFIG_KEY: 'request-log.output-dir',
  REQUEST_LOG_FLUSHINTERVAL_CONFIG_KEY: 'request-log.flushinterval',
  HEARTBEAT_PATH_CONFIG_KEY: 'health.heartbeat-path',
  INACTIVE_DURATION_SEC: 'sessions.inactiveDurationSec',
  MAX_SESSION_COUNT: 'sessions.maxSessionCount',
  ANALYTICSKEY_CONFIG_KEY: 'analytics.apiKey',
  CONTENT_BASEURL_KEY: 'content.baseUrl',
  PLATFORM_BASEURL_KEY: 'platform.baseUrl',
  CONFIG_NAME_UI: 'ui',
  CONFIG_NAME_UI_KEY: 'configs.ui',
  CONFIGS_FILE_PATH: 'configs-file-path',
  NFS_MOUNT_PATH: 'nfs-mount-path',
  NFS_MOUNT_PATH_DC: 'nfs-mount-path-dc',
  ANALYTICS_INTERNAL_USER: 'Internal',
  ANALYTICS_EXTERNAL_USER: 'External',
  ITS_API_URL: 'configs.its.apiUrl',
  ITS_BASE_URL: 'configs.its.baseUrl',
  DM_API_URL: 'configs.dm.apiUrl',
  DM_BASE_URL: 'configs.dm.baseUrl',
  ANSWERS_BASE_URL: 'configs.answers.baseUrl',
  ANSWERS_API_KEY: 'configs.answers.clientApiKey',
  UNIFIED_PORTAL_URL: 'configs.ui.unifiedPortalUrl',
  IDT_DOC_PORTAL_URL: 'configs.ui.docPortalUrl',
  CONTACT_SUPPORT_URL_EXTERNAL: 'configs.ui.contactSupportUrlExternal',
  CONTACT_SUPPORT_URL_INTERNAL: 'configs.ui.contactSupportUrlInternal',
  GCP_BUCKET_NAME: 'gcp.bucketName',
  DOWNLOAD_BASE_PATH: 'downloadBasePath',
  UPLOAD_BASE_PATH: 'uploadBasePath',
  COOKIES_EXPIRES: 'cookies.expires',
  REDIS_HOSTS: 'celery.redisHosts',
  REDIS_CLUSTER_NAME: 'celery.redisClusterName',
  REDIS_QUEUE_NAME: 'celery.queue',
  CELERY_TASK: 'worker.run', //The name of the python-celery task to run
  ENV: 'env',
  LOCALHOST: 'localhost',
  FETCH_CLIENT_PRODUCT: 'configs.its.fetchClientForProduct',
  FETCH_CLIENT_ANSWERS: 'configs.its.fetchClientForAnswers',
  FETCH_CLIENT_CONVERSATIONS: 'configs.its.fetchClientForConversations',
  PRODUCT_ITS: 'configs.its.product_its',
  LIVE_DATA_ANALYSIS: 'Conversations Analysis',
  SYSTEM_PROJ_DESCRIPTION: (data) => `${data.startDate} to ${data.endDate}`,
  LIVE_DATA_IN_REVIEW: 'Conversations Analysis*',
  CANDIDATE_FAQ: 'Candidate FAQ',

  /** Errors */
  ERRORS: {
    NODE_NAME_NOT_IN_NOUN_VERB_FORM: 'NODE_NAME_NOT_IN_NOUN_VERB_FORM',
    DB_CONNECTION_ERROR: 'DB_CONNECTION_ERROR',
    PROJECT_ALREADY_EXISTS: 'PROJECT_ALREADY_EXISTS',
    CAA_ALREADY_EXISTS: 'CAA_ALREADY_EXISTS',
    RUN_ALREADY_EXISTS: 'RUN_ALREADY_EXISTS',
    PROJECT_NOT_FOUND: 'PROJECT_NOT_FOUND',
    RUN_NOT_FOUND: 'RUN_NOT_FOUND',
    CAA_DOES_NOT_EXISTS: 'CAA_DOES_NOT_EXIST',
    BODY_PARAMETERS_MISSING: 'BODY_PARAMETERS_MISSING',
    QUERY_BUILDER_DATA_TYPE_NOT_SUPPORTED: 'QUERY_BUILDER_DATA_TYPE_NOT_SUPPORTED',
    INVALID_POST_DATA_APP_NAME_NOT_PROVIDED: 'INVALID_POST_DATA_APP_NAME_NOT_PROVIDED',
    INVALID_POST_DATA_DATASET_NAME_NOT_PROVIDED: 'INVALID_POST_DATA_DATASET_NAME_NOT_PROVIDED',
    INVALID_POST_DATA_FILE_NOT_PROVIDED: 'INVALID_POST_DATA_FILE_NOT_PROVIDED',
    INVALID_POST_DATA_RUN_NAME_NOT_PROVIDED: 'INVALID_POST_DATA_RUN_NAME_NOT_PROVIDED',
    INVALID_POST_DATA_NUM_OF_CLUSTERS_NOT_PROVIDED: 'INVALID_POST_DATA_NUM_OF_CLUSTERS_NOT_PROVIDED',
    INVALID_POST_DATA_NUM_OF_TURNS_NOT_PROVIDED: 'INVALID_POST_DATA_NUM_OF_TURNS_NOT_PROVIDED',
    RUN_PROCESS_NOT_COMPLETE: 'RUN_PROCESS_NOT_COMPLETE',
    INVALID_S3_LOCATION_URI: 'INVALID_S3_LOCATION_URI',
    DATASET_FILE_IS_NOT_CSV: 'DATASET_FILE_IS_NOT_CSV',
    DATASET_FILE_DOES_NOT_HAVE_FOUR_COLUMNS: 'DATASET_FILE_DOES_NOT_HAVE_FOUR_COLUMNS',
    FAIL_TO_DELETE_DATASET_FILE_FROM_LOCAL: 'FAIL_TO_DELETE_DATASET_FILE_FROM_LOCAL',
    FAIL_TO_CREATE_RUN_TASK: 'FAIL_TO_CREATE_RUN_TASK',
    FAIL_TO_CREATE_RUN_TASK_TRANS_MISSING: 'FAIL_TO_CREATE_RUN_TASK_TRANS_MISSING',
    NFS_UPLOAD_FILE_FAILED: 'NFS_UPLOAD_FILE_FAILED',
    NFS_DOWNLOAD_FILE_FAILED: 'NFS_DOWNLOAD_FILE_FAILED',
    NFS_DELETE_OBJECT_FAILED: 'NFS_DELETE_OBJECT_FAILED',
    NOT_A_NUMBER: 'NOT_A_NUMBER',
    FAIL_TO_UPDATE_ASSIGNED_INTENT: 'FAIL_TO_UPDATE_ASSIGNED_INTENT',
    FAIL_TO_INGEST_DATA: 'FAIL_TO_INGEST_DATA',
    FAILURE_DUE_TO_NON_ASCII_CHARACTERS_IN_DATASET: 'FAILURE_DUE_TO_NON_ASCII_CHARACTERS_IN_DATASET',
    MAX_PROJ_NUM_EXCEEDED: 'MAX_PROJ_NUM_EXCEEDED',
    MAX_RUN_NUM_EXCEEDED: 'MAX_RUN_NUM_EXCEEDED',
    DOWNLOAD_PERMISSION_DENIED: 'DOWNLOAD_PERMISSION_DENIED',
    CREATE_PERMISSION_DENIED: 'CREATE_PERMISSION_DENIED',
    UPDATE_PERMISSION_DENIED: 'UPDATE_PERMISSION_DENIED',
    CLIENT_ID_NOT_PROVIDED: 'CLIENT_ID_NOT_PROVIDED',
    LANGUAGE_ID_NOT_PROVIDED: 'LANGUAGE_ID_NOT_PROVIDED',
    INVALID_OR_EMPTY_UTTERANCES: 'INVALID_OR_EMPTY_UTTERANCES',
    RESPONSE_TITLE_OR_RESPONSE_CONTENT_NOT_PROVIDED: 'RESPONSE_TITLE_OR_RESPONSE_CONTENT_NOT_PROVIDED',
    RESPONSE_TITLE_OR_RESPONSE_ID_NOT_PROVIDED: 'RESPONSE_TITLE_OR_RESPONSE_ID_NOT_PROVIDED',
    CANDIDATE_FOLDER_NOT_CREATED: 'CANDIDATE_FOLDER_NOT_CREATED',
    API_KEY_NOT_PROVIDED: 'API_KEY_NOT_PROVIDED',
    FORBIDDEN: 'FORBIDDEN',
  },

  /**
   * Session and cookie info
   * @type {string}
   */
  SESSION_COOKIE_NAME: 'conversation_discovery',
  SESSION_COOKIE_SECRET: 'keyboard',

  /**
   * MySQL related
   * @type {string}
   */
  MYSQL_HOST: 'mysql.host',
  MYSQL_USER: 'mysql.user',
  MYSQL_PASSWORD: 'mysql.password',
  MYSQL_PORT: 'mysql.port',
  MYSQL_DATABASE: 'mysql.database',
  MYSQL_CONNECTION_LIMIT: 'mysql.connection_limit',

  /**
   * ES related
   * @type {string}
   */
  ES_NODE: 'es.nodes',
  ES_MAXRETRIES: 'es.max_retries',
  ES_REQUEST_TIMEOUT: 'es.request_timeout',
  ES_SNIFF_ON_START: 'es.sniff_on_start',
  ES_CLUSTER_INDEX: 'es.idt_cluster_index',
  ES_CONVERSATION_INDEX: 'es.idt_conversation_index',
  INGEST_BATCH_SIZE: 200,


  /**
     * CSV related
     * @type {string}
   */
  CONVERSATION_CSV_HEADER: 'interaction_id,sequence,turn,processed_body',

  /** DB ERROR CODES */
  DB_ERROR_CODES: {
    ER_NO_REFERENCED_ROW: 'ER_NO_REFERENCED_ROW',
    ER_NO_REFERENCED_ROW_2: 'ER_NO_REFERENCED_ROW_2',
    ER_DUP_ENTRY: 'ER_DUP_ENTRY'
  },

  //TODO: Have a map of code and description
  /** PROJECT STATUS CODES */
  PROJECT_STATUS_CODES: {
    NEW: 'NEW',
    IN_REVIEW: 'REVIEW',
    QUEUED: 'QUEUED',
    QUEUED_STATUS_DESCRIPTION: 'The project is queued',
    ENABLED: 'ENABLED',
    DISABLED: 'DISABLED',
    VERIFYING: 'VERIFYING',
    FAILED: 'FAILED'
  },

  PROJECT_TYPES: {
    MANUAL: 'MANUAL',
    SYSTEM: 'SYSTEM'
  },

  /** RUN STATUS CODES */
  RUN_STATUS_CODES: {
    QUEUED: 'QUEUED',
    READY: 'READY',
    COMPLETE: 'COMPLETE',
    QUEUED_STATUS_DESCRIPTION: 'The run is queued'
  },

  /** RUN DEFAULTS */
  RUN_DEFAULTS: {
    DEFAULT_RUN_NAME: 'DefaultAnalysis',
    DEFAULT_RUN_DESCRIPTION: 'This is a default analysis',
    DEFAULT_NUM_OF_CLUSTERS: -1,
    DEFAULT_NUM_OF_TURNS: 1,
    DEFAULT_STARRED_VALUE: 0,
    DEFAULT_STOP_WORDS: []
  },

  /** LIMITS */
  PROJ_LIMIT: 20,
  RUN_LIMIT: 20,

  /**
   * Security
   * @type {string}
   */
  IS_SECURITY_ENABLED: 'security.enabled',
  ISSUER: 'security.oidc.issuer',
  CLIENT_ID: 'security.oidc.client_id',
  CLIENT_SECRET: 'security.oidc.client_secret',
  REDIRECT_URI: 'security.oidc.redirect_uri',
  SCOPE: 'security.oidc.scope',
  OKTA_URL: 'security.okta_url',
  OKTA_ACCOUNT_URL: 'security.okta_account_url', // user Account url link
  AUTHORIZATION_ENABLED: 'security.authorization_enabled', //For tests, this will be false
  TRUSTED_FLOW_SEC_MESSAGE: 'UPDATE_SUCCESSFULL',

  /**
   * Project Parameters to be Updated
   */
  UPDATE_PROJECT_PARAMETERS: {
    PROJECT_NAME: 'projectName',
    PROJECT_DESCRIPTION: 'projectDescription'
  },

  TABLES: {
    PROJECT: 'project',
    RUN: 'run',
    CAA: 'caa'
  },

  HTTP_ERROR: {
    USER_IS_UNAUTHORIZED: 'USER_IS_UNAUTHORIZED',
    INTERNAL_SERVER_ERROR: 'INTERNAL_SERVER_ERROR'
  },

  PII_MASK_REGEX: {
    MASKED_EMAILID: 'abc@xyz.com',
    MASKED_DIGITS: '#',
    MASKED_SCRIPT_TAGS: '',
  },

  ANSWERS_API_URLS: {
    UPDATE_QUESTIONS: (data) => `${data.basePath}responses/addQuestionByResponseID/interfaceID/${data.interfaceId}/responseID/${data.responseId}?clientName=${data.clientId}`,
    POST_CANDIDATE_FAQ_FOLDER: (data) => `${data.basePath}folders/createFolder/interfaceID/${data.interfaceId}/parentFolderID/${data.parentFolderId}?clientName=${data.clientId}`,
    GET_INTERFACES_LIST: (data) => `${data.basePath}interface/listInterface?clientName=${data.clientId}`,
    GET_FAQS: (data) => `${data.basePath}responses/listResponsesByFolder/interfaceID/${data.interfaceId}/${data.folderId}?clientName=${data.clientId}`,
    POST_FAQ: (data) => `${data.basePath}responses/createResponseContentByFolder/interfaceID/${data.interfaceId}/folderID/${data.folderId}/languageID/${data.languageId}?clientName=${data.clientId}`
  },

  API_URLS: {
    GET_INTENTS: (data) => `${data.basePath}clients/${data.clientId}/apps/${data.appId}/nodes?offset=${data.offset}&limit=${data.limit}&isModified=${data.isModified}`,
    UPDATE_INTENTS: (data) => `${data.basePath}clients/${data.clientId}/apps/${data.appId}/nodes/${data.nodeId}`,
    CREATE_INTENTS: (data) => `${data.basePath}clients/${data.clientId}/apps/${data.appId}/nodes/${data.parentNodeId}/children`,
    GET_CLIENT_CONFIG: (data) => `${data.itsApiUrl}clients?includeApps=true`,
  },
  INTENT_FETCH_LIMIT: 200,
  DELETE_STATUS_UPDATE: {
    starred: 1
  },
  RETRY_THRESHOLD: 3,

  roles: {
    unauthorized: 'unauthorized',
    viewer: 'viewer',
    tester: 'tester',
    developer: 'developer',
    operator: 'operator',
    admin: 'admin',
    internaladmin: 'internaladmin'
  },
  
  roleHierarchy: {
    ['unauthorized']: 0,
    ['viewer']: 1,
    ['tester']: 2,
    ['developer']: 3,
    ['operator']: 4,
    ['admin']: 5,
    ['internaladmin']: 6
  },
  
  viewer: ['GET'],
};
