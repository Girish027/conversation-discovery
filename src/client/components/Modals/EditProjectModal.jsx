import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Dialog } from '@tfs/ui-components';
import constants from '../../constants';
import Form from '../Form/Form';
import {
  projectActions,
} from '../../state/projectsState';
import {
  dialogStyles
} from '../../styles';

import editRunUiSchema from '../../schema/editProject/uiSchema.json';
import editRunJsonSchema from '../../schema/editProject/jsonSchema.json';
import ErrorLayout from './ErrorLayout';

export default class EditProjectModal extends Component {
  constructor(props) {
    super(props);
    this.form = React.createRef();
    this.props = props;

    this.onClickSubmit = this.onClickSubmit.bind(this);
    this.onClickCancel = this.onClickCancel.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.handleOnchange = this.handleOnchange.bind(this);

    this.styleOverride = Object.assign({}, {
      ...dialogStyles,
      container: {
        width: '450px',
        height: '410px',
        display: 'grid',
        gridTemplateRows: '60px auto 60px',
      },
      childContainer: {
        marginTop: '10px',
        marginBottom: '10px',
      },
      content: {
        maxWidth: '450px',
        maxHeight: '410px',
        left: 'calc((100vw - 450px) / 2)'
      }
    },
    this.props.styleOverride
    );

    this.state = {
      formData: {},
      isValid: false
    };
  }

  onClickSubmit() {
    this.form.current.submit();
  }

  onClickCancel() {
    const { onClickClose } = this.props;

    onClickClose();
  }

  onSubmit() {
    const { dispatch, projectId } = this.props;
    const { formData } = this.state;
    const { projectName = '', projectDescription = '' } = formData;
    dispatch(projectActions.editProject(projectId, { projectName, projectDescription }));
  }

  handleOnchange(data, errors) {
    if (errors.length === 0) {
      this.setState({
        formData: data,
        isValid: true
      });
    } else if (errors.length > 0) {
      this.setState({
        isValid: false
      });
    }
  }

  render() {
    const {
      header, size, formData, errorType, ...otherProps
    } = this.props;

    const { isValid } = this.state;

    const { buttons } = constants;
    const { submit: okChildren } = buttons;
    const { cancel: cancelChildren } = buttons;

    return (
      <div>
        <Dialog
          size={size}
          headerChildren={header}
          {...otherProps}
          isOpen
          okVisible
          okDisabled={!isValid}
          okChildren={okChildren}
          cancelVisible
          cancelChildren={cancelChildren}
          closeIconVisible
          styleOverride={this.styleOverride}
          onClickOk={this.onClickSubmit}
          onClickCancel={this.onClickCancel}
          centerContent={false}
        >
          <div>
            {errorType && (
              <ErrorLayout errorMsg={errorType} />
            )}
            <Form
              uiSchema={editRunUiSchema}
              jsonSchema={editRunJsonSchema}
              onSubmit={this.onSubmit}
              onChange={this.handleOnchange}
              formData={formData}
              ref={this.form}
              liveValidate
            />
          </div>
        </Dialog>
      </div>
    );
  }
}

EditProjectModal.propTypes = {
  dispatch: PropTypes.func,
  header: PropTypes.string,
  projectId: PropTypes.string.isRequired,
  closeIconVisible: PropTypes.bool,
  onClickOk: PropTypes.func,
  onClickClose: PropTypes.func,
  styleOverride: PropTypes.object,
  size: PropTypes.string,
  formData: PropTypes.object,
  errorType: PropTypes.string,
};

EditProjectModal.defaultProps = {
  dispatch: () => {},
  closeIconVisible: true,
  onClickOk: () => {},
  onClickClose: () => {},
  styleOverride: {},
  size: 'small',
  errorType: '',
  header: constants.modalInfo.editProject.header,
};
