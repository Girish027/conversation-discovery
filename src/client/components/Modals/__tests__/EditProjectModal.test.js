import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import EditProjectModal from '../EditProjectModal';
import constants from '../../../constants';
import {
  projectActions,
} from '../../../state/projectsState';

describe('<EditProjectModal />', function () {
  let wrapper;

  const props = {
    projectId: 'project-123',
    header: constants.modalInfo.editProject.header,
    formData: {
      projectName: 'Test Project',
      projectDescription: 'This is a test project',
    }
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  const getShallowWrapper = (propsObj) => shallow(<EditProjectModal
    {...propsObj}
  />);

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getShallowWrapper(props);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly with default props', () => {
      wrapper = getShallowWrapper(props);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality', function () {
    beforeEach(function () {
      props.dispatch = jest.fn();
      props.onClickClose = jest.fn();
      projectActions.editProject = jest.fn(() => 'editProject');
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('onSubmit', () => {
      const newData = {
        projectName: 'new project name',
        projectDescription: 'new project desc'
      };
      wrapper = getShallowWrapper(props);
      wrapper.setState({
        formData: newData
      });
      wrapper.instance().onSubmit();
      expect(projectActions.editProject).toHaveBeenCalledWith(props.projectId, newData);
      expect(props.dispatch).toHaveBeenCalledWith('editProject');
    });

    test('onClickCancel', () => {
      wrapper = getShallowWrapper(props);
      wrapper.instance().onClickCancel();
      expect(props.onClickClose).toHaveBeenCalled();
    });

    test('handleOnchange when there is no error', () => {
      wrapper = getShallowWrapper(props);
      const newData = {
        projectName: 'new project name',
        projectDescription: 'new project desc'
      };
      wrapper.instance().handleOnchange(newData, []);
      expect(wrapper.state().formData).toBe(newData);
      expect(wrapper.state().isValid).toBe(true);
    });

    test('handleOnchange when there is error', () => {
      wrapper = getShallowWrapper(props);
      const newData = {
        projectName: 'new project name',
        projectDescription: 'new project desc'
      };
      wrapper.instance().handleOnchange(newData, ['duplicate project']);
      expect(wrapper.state().isValid).toBe(false);
    });
  });
});
