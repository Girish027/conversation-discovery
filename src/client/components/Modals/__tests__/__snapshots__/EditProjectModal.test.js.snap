// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditProjectModal /> Snapshots renders correctly with default props 1`] = `
<div>
  <withTheme(DialogComponent)
    cancelChildren="CANCEL"
    cancelVisible={true}
    centerContent={false}
    closeIconVisible={true}
    dispatch={[Function]}
    headerChildren="Edit Transcript Details"
    isOpen={true}
    okChildren="SUBMIT"
    okDisabled={true}
    okVisible={true}
    onClickCancel={[Function]}
    onClickClose={[Function]}
    onClickOk={[Function]}
    projectId="project-123"
    size="small"
    styleOverride={
      Object {
        "cancel": Object {
          "backgroundColor": "unset",
          "paddingLeft": "10px",
          "paddingRight": "10px",
        },
        "childContainer": Object {
          "marginBottom": "10px",
          "marginTop": "10px",
        },
        "container": Object {
          "display": "grid",
          "gridTemplateRows": "60px auto 60px",
          "height": "410px",
          "width": "450px",
        },
        "content": Object {
          "left": "calc((100vw - 450px) / 2)",
          "maxHeight": "410px",
          "maxWidth": "450px",
        },
        "ok": Object {
          "marginLeft": "10px",
          "paddingLeft": "25px",
          "paddingRight": "25px",
        },
        "top": "20%",
      }
    }
  >
    <div>
      <Form
        formContext={Object {}}
        formData={
          Object {
            "projectDescription": "This is a test project",
            "projectName": "Test Project",
          }
        }
        jsonSchema={
          Object {
            "properties": Object {
              "projectDescription": Object {
                "maxLength": 150,
                "minLength": 0,
                "placeholder": "Add Project Description",
                "title": "Description",
                "type": "string",
              },
              "projectName": Object {
                "maxLength": 50,
                "minLength": 4,
                "placeholder": "Add Unique Project Name",
                "title": "Name",
                "type": "string",
              },
            },
            "title": "",
            "type": "object",
          }
        }
        liveValidate={true}
        onChange={[Function]}
        onSubmit={[Function]}
        showErrorList={false}
        uiSchema={
          Object {
            "projectDescription": Object {
              "ui:field": "textarea",
            },
            "projectName": Object {
              "ui:field": "text",
            },
          }
        }
      />
    </div>
  </withTheme(DialogComponent)>
</div>
`;
